File: Algorithms/src/85. Maximal Rectangle/Code/solution.js
var maximalRectangle = function(matrix) {
    if (matrix.length === 0) return 0;

    const cols = matrix[0].length;
    let heights = new Array(cols).fill(0);
    let maxArea = 0;

    const largestRectangleArea = (heights) => {
        let stack = [];
        let max = 0;
        heights.push(0);

        for (let i = 0; i < heights.length; i++) {
            while (stack.length && heights[stack[stack.length - 1]] > heights[i]) {
                let h = heights[stack.pop()];
                let w = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
                max = Math.max(max, h * w);
            }
            stack.push(i);
        }
        heights.pop();
        return max;
    };

    for (let row of matrix) {
        for (let j = 0; j < cols; j++) {
            heights[j] = row[j] === '1' ? heights[j] + 1 : 0;
        }
        maxArea = Math.max(maxArea, largestRectangleArea(heights));
    }

    return maxArea;
};

File: Algorithms/src/85. Maximal Rectangle/Code/solution.java
class Solution {
    private int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0;
        int n = heights.length;
        int[] arr = new int[n + 1];
        System.arraycopy(heights, 0, arr, 0, n);

        for (int i = 0; i <= n; i++) {
            while (!stack.isEmpty() && arr[stack.peek()] > arr[i]) {
                int h = arr[stack.pop()];
                int w = stack.isEmpty() ? i : i - stack.peek() - 1;
                maxArea = Math.max(maxArea, h * w);
            }
            stack.push(i);
        }
        return maxArea;
    }

    public int maximalRectangle(char[][] matrix) {
        if (matrix.length == 0)
            return 0;

        int cols = matrix[0].length;
        int[] heights = new int[cols];
        int ans = 0;

        for (char[] row : matrix) {
            for (int j = 0; j < cols; j++) {
                heights[j] = row[j] == '1' ? heights[j] + 1 : 0;
            }
            ans = Math.max(ans, largestRectangleArea(heights));
        }
        return ans;
    }
}
